### DISCUSSION TITLE
How to better modify AST nodes?

### DISCUSSION THREAD

**Discussion by yisar** [swc-project/swc#2036] [Q&A] [answered]
What should I do to replace an AST node with another node of a different type?

For example:

```js
// input:
<For each={list}/>
// output:
{list.map()}
```

I find that SWC is difficult to modify when the semantics and types are completely different. Is there any good idea here?

cc @devongovett @kdy1

**Comment by kdy1** [comment, marked as answer]
Currently, swc is designed to modify AST nodes from rust side.

if you are using swc form rust, you can simply implement [Fold](https://docs.rs/swc_ecma_visit/0.35.1/swc_ecma_visit/trait.Fold.html).

**Reply by yisar** [reply]
Yes, it is rust. What's the difference between fold and visit_mut?

**Reply by kdy1** [reply]
It's documented at https://docs.rs/swc_visit/0.2.5/swc_visit/#fold-vs-visitmut

**Comment by yisar** [comment]
@kdy1 
```rust
fn fold_jsx_namespaced_name(&mut self, node: ast::JSXNamespacedName) -> ast::Ident { // here renturn a different type
        if *node.name.sym == *"key" {
            return ast::Ident {
                span: DUMMY_SP,
                optional: false,
                sym: format!("{}", "key").into(),
            };
        } else {
            return node;
        }
 }
```

In fact, I want to return a completely different ast node. I need to completely distort the semantics and modify the structure in the right place.

I really can't find a good way. I've cried many times. Do you have a way?

**Reply by kdy1** [reply]
You can override `fold_expr` instead. Be sure to handle children nodes using `node.fold_children_with(self)` if required.

**Reply by kdy1** [reply]
One good example is the jsx pass.

https://github.com/swc-project/swc/blob/2c52021ed42adc8d3524562c21a0dfca31c8d110/ecmascript/transforms/react/src/jsx/mod.rs#L884-L917

**Comment by yisar** [comment]
I tried your method and found it still couldn't work.
I want to input `Expr` and output `Stmt`.

```js
// input:
<For each={list}/>
// output:
{list.map()} // this is a Stmt, not Expr.
```
not this:
```js
// input:
<For each={list}/>
// output:
list.map()
```
Do you have any ideas in this situation?

@kdy1

**Reply by kdy1** [reply]
Implement `visit_mut_stmt`

**Reply by yisar** [reply]
> Implement visit_mut_stmt

```<For each={list}/>``` is not a Stmt, but a Expr, it cound not enter the ```visit_mut_stmt``` function.

**Reply by kdy1** [reply]
No, every expression belong to a statement.
It's`ExpressionStatement`, not `Expression`.

**Reply by kdy1** [reply]
In swc, it's `Stmt::Expr` or `ExprStmt`

**Reply by yisar** [reply]
> In swc, it's `Stmt::Expr` or `ExprStmt`

Thank u, i am going to have a try!

**Comment by yisar** [comment]
@kdy1 Thank you, I found a better way to distort semantics, use `JSXElementChild`

```rs
match c {
            ast::JSXElementChild::JSXElement(v) => {
                ast::JSXElementChild::JSXExprContainer...
```
I have to say, it's really hard [cring]

**Comment by yisar** [comment]
@kdy1 I am comming back 😭 

```js
// input
<Component>
  <view slot=header>header</view>
</Component>

// output
<Component $slot$header={<view>header</view>}>
</Component>
```
When I need to change the slot tag into render props, I need to modify the reference of the parent. Do you have a good idea?

**Reply by kdy1** [reply]
If you need to reference the value you are going to modify, you have to clone it.

**Reply by yisar** [reply]
> If you need to reference the value you are going to modify, you have to clone it.

No, I can't. I need to modify the AST, but I can't clone it.

Or is there another way? Have you ever written similar use cases?

**Reply by kdy1** [reply]
There are some cases which requires info from parent, but if you want to use the parent node, you have to clone it.
Is there any reason you can't clone?

**Reply by yisar** [reply]
I need to use the child information, and then modify the parent node. 
Look at the above use cases carefully. The child will become the parent's props

**Reply by kdy1** [reply]
Then you can implement the visitor method for the parent node.

**Reply by yisar** [reply]
Do you mean to traverse twice? First, traverse the children to collect information, and then modify the value of the parent through the visitor?

**Reply by kdy1** [reply]
I didn't mean it, but it's also a good solution.
